### DISCUSSION TITLE
How crucial is Spire for supporting BigDecimal?

### DISCUSSION THREAD

**Discussion by benhutchison** [erikerlandson/coulomb#441] [General]
@erikerlandson In the docs, you say:

_You may notice that BigInt and BigDecimal are not supported in coulomb-core, even though they are Scala native types. This is because Scala's native numeric typeclasses do not uniformly treat the core types Int, Long, Float, Double the same as BigInt and BigDecimal, while spire's typeclasses do. For this reason, it is much easier to support BigInt and BigDecimal as part of the coulomb-spire package._

Can you expand on why Spire makes supporting `BigDecimal` easier? Or flipped, what would the obstacles be to supporting BigDecimal without Spire?

The issue is that decimals are the standard type for any kind of monetary or financial calculation. That's a common, widespread application domain. Meanwhile, Spire is quite a heavy and complex dependency to bring in, and requiring it for any user of Coulomb who needs to work with money seems a little unfortunate. 

I'm wondering if there are ways to peel BigDecimal and Spire apart..?

**Comment by erikerlandson** [comment]
It has to do with `BigInt` and `BigDecimal` not being members of `Numeric` or `Fractional`:
https://github.com/erikerlandson/coulomb/blob/scala3/core/src/main/scala/coulomb/conversion/standard/value.scala

It's not conceptually hard, it's just that I'd have to write the conversions myself, and the number of them grows quadratically with the number of types you are supporting.

Spire IS heavy-weight, which is one reason I elected to not make it a dependency of `coulomb-core`, but one thing they they did get right was a unified conversion typeclass for ALL the numeric types (theirs and scala's):
https://github.com/erikerlandson/coulomb/blob/scala3/spire/src/main/scala/coulomb/conversion/spire/value.scala

The other annoyance is that since scala Numeric and spire's conversions both overlap, but are incompatible, I could not write an "overlay" policy - you need to either import the spire conversions or core conversions, but importing both will break.  On the bright side, spire's support is a superset of core support, so if you want spire it's an easy choice to make, just import the spire policies.

If you wanted to write the support for BigDecimal and BigInt into the `standard` policy, that's a pull request I'd probably accept :grin:

**Reply by erikerlandson** [reply]
oh, you'd also need to add them to the unit conversions and the value promotion policy, those are pretty easy

**Reply by erikerlandson** [reply]
I changed the interface of promotion policies a bit in my parsing PR, to use scala's `*:` instead of `&:`
